sympy/simplify/cse_main.py

```diff
@@ -536,3 +536,101 @@ def cse(exprs, symbols=None, optimizations=None, postprocess=None,
         return replacements, reduced_exprs
 
     return postprocess(replacements, reduced_exprs)
+
+
+class Marker:
+    pass
+
+# XXX: Should this go in sympy.utilities.iterables?
+def shortest_repeated_subsequence(S, ignore=(Marker,)):
```

Comment:
@smichr you've done a lot of work in the iterables and combinatorics module. Any thoughts on this? 

My planned cse algorithm is roughly: 
  
   1. Concatenate all nc-Mul args with a marker
   2. Find a shortest repeated subsequence (i.e., sub-product)
   3. Replace it everywhere with a cse dummy symbol
   4. Repeat until repeated subsequences are found